File: 7-kyu-kata/name-that-number/python/name_that_number_test.py
# 7 kyu - Name That Number!  [ ID: 579ba41ce298a73aaa000255  (name-that-number) ]
# URL: https://www.codewars.com/kata/579ba41ce298a73aaa000255
# Category: REFERENCE  |  Tags: FUNDAMENTALS
# ******************************************************************************
from random import randint, shuffle

def solution(x):
    numbers = ["zero", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten"]
    ten_mults = ["twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"]
    teens = ["ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"]
    
    if x % 100 < 10:
        return numbers[x % 100]
    elif x % 100 < 20:
        return teens[(x % 100) - 10]
    else:
        if (x % 100) % 10 == 0:
            return ten_mults[(x // 10) % 10 - 2]
        else:
            return ten_mults[(x // 10) % 10 - 2] + ' ' + numbers[x % 10]



test.describe('Basic Tests')
test.assert_equals(name_that_number(1), 'one')
test.assert_equals(name_that_number(52), 'fifty two')
test.assert_equals(name_that_number(21), 'twenty one')
test.assert_equals(name_that_number(99), 'ninety nine')
test.assert_equals(name_that_number(0), 'zero')
test.assert_equals(name_that_number(53), 'fifty three')
test.assert_equals(name_that_number(23), 'twenty three')
test.assert_equals(name_that_number(76), 'seventy six')


test.describe('Random Tests')
lst = list(range(100))
shuffle(lst)
for n in lst:
    test.it(repr(n))
    test.assert_equals(name_that_number(n), solution(n))

File: kata-7-kyu/name-that-number/javascript/nameThatNumber.Test.js
//+ ====================================================================================================================
//+
//+ 7 kyu - Name That Number!  [ ID: 579ba41ce298a73aaa000255 ] (name-that-number)
//+ URL: https://www.codewars.com/kata/579ba41ce298a73aaa000255
//+ Category: REFERENCE  |  Tags: FUNDAMENTALS
//+
//+ ====================================================================================================================

const { assert } = require("chai")
const { nameThatNumber } = require("./nameThatNumber")

const words = [
  [90, "ninety"],
  [80, "eighty"],
  [70, "seventy"],
  [60, "sixty"],
  [50, "fifty"],
  [40, "forty"],
  [30, "thirty"],
  [20, "twenty"],
  [19, "nineteen"],
  [18, "eighteen"],
  [17, "seventeen"],
  [16, "sixteen"],
  [15, "fifteen"],
  [14, "fourteen"],
  [13, "thirteen"],
  [12, "twelve"],
  [11, "eleven"],
  [10, "ten"],
  [9, "nine"],
  [8, "eight"],
  [7, "seven"],
  [6, "six"],
  [5, "five"],
  [4, "four"],
  [3, "three"],
  [2, "two"],
  [1, "one"]
]

function solution(num) {
  let current = null
  let quo = null
  let wordValue = null
  let result = []
  for (let i = 0; i < words.length; i++) {
    current = words[i]
    wordValue = current[0]
    quo = Math.floor(num / wordValue)
    num %= wordValue
    if (quo) result.push(current[1])
  }
  return result.join(" ") || "zero"
}

describe("Basic Tests", function () {
  it("1", function () {
    assert.strictEqual(nameThatNumber(1), "one")
  })
  it("52", function () {
    assert.strictEqual(nameThatNumber(52), "fifty two")
  })
  it("21", function () {
    assert.strictEqual(nameThatNumber(21), "twenty one")
  })
  it("99", function () {
    assert.strictEqual(nameThatNumber(99), "ninety nine")
  })
  it("0", function () {
    assert.strictEqual(nameThatNumber(0), "zero")
  })
  it("53", function () {
    assert.strictEqual(nameThatNumber(53), "fifty three")
  })
  it("23", function () {
    assert.strictEqual(nameThatNumber(23), "twenty three")
  })
  it("76", function () {
    assert.strictEqual(nameThatNumber(76), "seventy six")
  })
})

describe("Random Tests", function () {
  const shuffleArray = (array) => {
    for (let i = array.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1))
      const temp = array[i]
      array[i] = array[j]
      array[j] = temp
    }
  }

  const arr = [...Array(100).keys()]

  shuffleArray(arr)

  for (let i of arr) {
    it(`${i}`, function () {
      assert.strictEqual(nameThatNumber(i), solution(i))
    })
  }
})
